### PR TITLE
Add `-Zerror-metrics=PATH` to save diagnostic metadata to disk

### PR DESCRIPTION
**Pull request by estebank**
Add nightly only flag to store information about encountered errors to disk. The information being tracked is:

 - Error code
 - Path to error being emitted
 - Number of subdiagnostics
 - Number of structured suggestions

### CODE REVIEW

File: compiler/rustc_session/src/session.rs [hunk 1]
```diff
@@ -1009,6 +1010,12 @@ fn default_emitter(
                 );
                 Box::new(emitter.ui_testing(sopts.unstable_opts.ui_testing))
             } else {
+                let metrics = sopts.unstable_opts.error_metrics.as_ref().and_then(|path| {
+                    let mut path = path.clone();
+                    std::fs::create_dir_all(&path).ok()?;
+                    path.push(format!("error_metrics_{}", std::process::id()));
```
Comment (compiler-errors) [inline compiler/rustc_session/src/session.rs:1016]: Why PID? Can we use something more ordered such as `YYYY-MM-DD-HH-SS-PID`?

**Comment by Mark-Simulacrum** [comment]
Does this need to be in rustc? What is the intended consumption workflow (is there intent to stabilize this)? I would expect that ~all of this information is already available to programs running rustc with errors emitted as JSON (e.g., Cargo) and they could handle aggregating and propagating that information.

**Comment by Dylan-DPC** [comment]
@estebank any updates on this? thanks

File: compiler/rustc_session/src/session.rs (hunk 1 above)
Comment (yaahc) [reply compiler/rustc_session/src/session.rs:1016]: I am not sure of the precise reasoning behind using PID, but I have no objections to switching to datetime-pid. We can definitely use that instead.

**Comment by yaahc** [comment]
Hey y'all. I'm going to be picking up some of this work and wanted to give a quick update on the next steps I'm planning and respond to the questions in the thread. 

> Does this need to be in rustc? ... I would expect that ~all of this information is already available to programs running rustc with errors emitted as JSON (e.g., Cargo) and they could handle aggregating and propagating that information.

@Mark-Simulacrum, the specific metrics included in this PR, such as the counts of specific error codes and subdiagnostics, definitely don't need to live in rustc, and we have no objection to moving those downstream to the consumers of the JSON output. There will still need to be some machinery in rustc for metrics we care about that don't make sense to emit to cargo or other consumers, such as saving ICE reports or logging for events that we might turn into assertions in the future (e.g. incremental compilation and situations like https://blog.rust-lang.org/2021/05/10/Rust-1.52.1.html.

> What is the intended consumption workflow 

The plans for consumption workflows are mostly forming at the moment; we've thrown around ideas like having cargo plugins to summarize information, asking users to grep metrics for specific events on nightly, building support within crater/perf to search for issues in the metrics output, or having rust-analyzer look for issues and notify users when they've encountered a problem we'd like reported. I imagine the direction we go will depend heavily on the issues that we first start leveraging the metrics to track.

> (is there intent to stabilize this)? 

The only part we anticipate wanting to stabilize is the flag itself, as per the MCP, which configures the directory where metrics should be stored.

## Plan going forward

I will close this PR and open a new one that adds the flag and uses it as the default location for saving ICE reports, assuming the `RUST_ICE` environment variable isn't set.

**Comment by Dylan-DPC** [comment]
Closing this as per https://github.com/rust-lang/rust/pull/119355#issuecomment-2261166160

**Pull request closed without merging**